### PR TITLE
Fwd loss

### PR DESCRIPTION
**Pull request by olliethomas**

### CODE REVIEW

**Comment by olliethomas** [comment]
Could potentially add a model that uses the vision datasets

File: ethicml/utility/lookup.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+from typing import Dict
+
+from ethicml.data import Dataset, adult, compas, credit, german, nonbinary_toy, sqf, toy
+
+__all__ = ["get_dataset_obj_by_name"]
+
+
+def get_dataset_obj_by_name(name: str) -> Dataset:
+    """Given a dataset name, get the corresponding dataset object."""
+    lookup: Dict[str, Dataset] = {
```
Comment (thomkeh) [inline ethicml/utility/lookup.py:12]:
```suggestion
    lookup: Dict[str, Callable[[], Dataset] = {
```

We could also make it so that we just look up the function and only then call it to get the dataset object.

Comment (olliethomas) [reply ethicml/utility/lookup.py:12]: I'm not sure as this isn't something that's used all that often. In fact, given the recent changes to Dataset, I'm not sure if this is still needed, but it probably falls out of the scope of this PR to check that

Comment (thomkeh) [reply ethicml/utility/lookup.py]: yeah it was probably a micro optimization

**Comment by thomkeh** [comment]
maybe mention "DRO" (distributionally robust optimization) somewhere?

File: ethicml/implementations/fwd_modules/fwd_loss.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+from torch import Tensor, nn
+from torch.nn.modules.loss import NLLLoss, _Loss
+
+from ethicml.common import implements
+
+__all__ = ["FWDLoss"]
+
+
+class FWDLoss(nn.Module):
```
Comment (MylesBartlett) [inline ethicml/implementations/fwd_modules/fwd_loss.py:12]:
```suggestion
class DROLoss(nn.Module):
```

File: ethicml/implementations/fwd_modules/fwd_loss.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+"""FWD Loss."""
+from typing import Optional
+
+from torch import Tensor, nn
+from torch.nn.modules.loss import NLLLoss, _Loss
+
+from ethicml.common import implements
+
+__all__ = ["FWDLoss"]
```
Comment (MylesBartlett) [inline ethicml/implementations/fwd_modules/fwd_loss.py:9]:
```suggestion
__all__ = ["DROLoss"]
```

File: ethicml/implementations/fwd_tabular.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+from torch.utils.data import DataLoader
+
+from ethicml.implementations.beutel import set_seed
+from ethicml.implementations.fwd_modules.fwd_classifier import FWDClassifier
+from ethicml.implementations.pytorch_common import CustomDataset, TestDataset
+from ethicml.implementations.utils import InAlgoArgs, load_data_from_flags
+from ethicml.utility import DataTuple, SoftPrediction, TestTuple
+
+
+class FwdArgs(InAlgoArgs):
```
Comment (MylesBartlett) [inline ethicml/implementations/fwd_tabular.py:18]:
```suggestion
class DroArgs(InAlgoArgs):
```

File: ethicml/implementations/fwd_modules/fwd_loss.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+class FWDLoss(nn.Module):
+    """Fairness Without Demographics Loss."""
+
+    def __init__(self, loss_module: Optional[_Loss] = None, eta: float = 0.5):
+        """Set up the loss, set which loss you want to optimize and the eta to offset by."""
+        super().__init__()
+        if loss_module is None:
+            loss_module = NLLLoss()
+        self.loss = loss_module
```
Comment (MylesBartlett) [inline ethicml/implementations/fwd_modules/fwd_loss.py:20]:
```suggestion
        else:
            assert loss_module.reduction == "none"
        self.loss = loss_module
```
